File: apps/legacy/apps/backend/src/routers/documents.py
from fastapi import APIRouter, UploadFile, File, Form, HTTPException, Query, Depends
from fastapi.responses import StreamingResponse, JSONResponse
from typing import Optional, List
import uuid
import os
from datetime import datetime, timedelta
import mimetypes

router = APIRouter()

# In-memory storage for demo (replace with database in production)
documents_db = {}

# Mock tenant context
def get_current_tenant(tenant_id: str = Query(...)):
    return tenant_id

@router.get("")
async def list_documents(
    parentId: Optional[str] = None,
    tenantId: str = Depends(get_current_tenant)
):
    """List documents in a folder"""
    docs = [
        doc for doc in documents_db.values()
        if doc["tenantId"] == tenantId 
        and doc.get("parentId") == parentId
        and doc.get("deletedAt") is None
    ]
    return docs

@router.post("/folder")
async def create_folder(
    name: str = Form(...),
    parentId: Optional[str] = Form(None),
    tenantId: str = Form(...)
):
    """Create a new folder"""
    doc_id = str(uuid.uuid4())
    now = datetime.utcnow().isoformat()
    
    # Check for duplicates
    for doc in documents_db.values():
        if (doc["tenantId"] == tenantId and 
            doc.get("parentId") == parentId and 
            doc["name"] == name and
            doc.get("deletedAt") is None):
            raise HTTPException(status_code=400, detail="Folder already exists")
    
    document = {
        "id": doc_id,
        "tenantId": tenantId,
        "parentId": parentId,
        "name": name,
        "type": "folder",
        "mimeType": None,
        "sizeBytes": None,
        "storagePath": None,
        "isShared": False,
        "sharedLinkToken": None,
        "sharedLinkExpiresAt": None,
        "createdBy": None,
        "createdAt": now,
        "updatedAt": now,
        "deletedAt": None
    }
    documents_db[doc_id] = document
    return document

@router.post("/upload")
async def upload_file(
    file: UploadFile = File(...),
    parentId: Optional[str] = Form(None),
    tenantId: str = Form(...)
):
    """Upload a file"""
    doc_id = str(uuid.uuid4())
    now = datetime.utcnow().isoformat()
    
    # Get file size
    content = await file.read()
    size_bytes = len(content)
    await file.seek(0)
    
    # Detect mime type
    mime_type = file.content_type or mimetypes.guess_type(file.filename)[0] or "application/octet-stream"
    
    # Check for duplicates
    for doc in documents_db.values():
        if (doc["tenantId"] == tenantId and 
            doc.get("parentId") == parentId and 
            doc["name"] == file.filename and
            doc.get("deletedAt") is None):
            raise HTTPException(status_code=400, detail="File already exists")
    
    document = {
        "id": doc_id,
        "tenantId": tenantId,
        "parentId": parentId,
        "name": file.filename,
        "type": "file",
        "mimeType": mime_type,
        "sizeBytes": size_bytes,
        "storagePath": f"{tenantId}/{doc_id}",
        "isShared": False,
        "sharedLinkToken": None,
        "sharedLinkExpiresAt": None,
        "createdBy": None,
        "createdAt": now,
        "updatedAt": now,
        "deletedAt": None,
        "content": content  # Store content in memory for demo
    }
    documents_db[doc_id] = document
    
    # Return without content
    response_doc = {k: v for k, v in document.items() if k != "content"}
    return response_doc

@router.get("/{document_id}/download")
async def download_file(
    document_id: str,
    tenantId: str = Query(...)
):
    """Download a file"""
    doc = documents_db.get(document_id)
    if not doc:
        raise HTTPException(status_code=404, detail="Document not found")
    if doc["tenantId"] != tenantId:
        raise HTTPException(status_code=403, detail="Access denied")
    if doc["type"] != "file":
        raise HTTPException(status_code=400, detail="Not a file")
    
    content = doc.get("content", b"")
    
    return StreamingResponse(
        iter([content]),
        media_type=doc["mimeType"] or "application/octet-stream",
        headers={"Content-Disposition": f'attachment; filename="{doc["name"]}"'}
    )

@router.get("/{document_id}/preview")
async def preview_file(
    document_id: str,
    tenantId: str = Query(...)
):
    """Get file preview URL"""
    doc = documents_db.get(document_id)
    if not doc:
        raise HTTPException(status_code=404, detail="Document not found")
    if doc["tenantId"] != tenantId:
        raise HTTPException(status_code=403, detail="Access denied")
    if doc["type"] != "file":
        raise HTTPException(status_code=400, detail="Not a file")
    
    # Return inline content for preview
    content = doc.get("content", b"")
    
    return StreamingResponse(
        iter([content]),
        media_type=doc["mimeType"] or "application/octet-stream",
        headers={"Content-Disposition": f'inline; filename="{doc["name"]}"'}
    )

@router.patch("/{document_id}")
async def rename_document(
    document_id: str,
    name: str = Form(...),
    tenantId: str = Form(...)
):
    """Rename a document"""
    doc = documents_db.get(document_id)
    if not doc:
        raise HTTPException(status_code=404, detail="Document not found")
    if doc["tenantId"] != tenantId:
        raise HTTPException(status_code=403, detail="Access denied")
    
    doc["name"] = name
    doc["updatedAt"] = datetime.utcnow().isoformat()
    return {k: v for k, v in doc.items() if k != "content"}

@router.patch("/{document_id}/move")
async def move_document(
    document_id: str,
    parentId: Optional[str] = Form(None),
    tenantId: str = Form(...)
):
    """Move a document to a different folder"""
    doc = documents_db.get(document_id)
    if not doc:
        raise HTTPException(status_code=404, detail="Document not found")
    if doc["tenantId"] != tenantId:
        raise HTTPException(status_code=403, detail="Access denied")
    
    doc["parentId"] = parentId
    doc["updatedAt"] = datetime.utcnow().isoformat()
    return {k: v for k, v in doc.items() if k != "content"}

@router.delete("/{document_id}")
async def delete_document(
    document_id: str,
    tenantId: str = Query(...),
    permanent: bool = Query(False)
):
    """Soft delete or permanently delete a document"""
    doc = documents_db.get(document_id)
    if not doc:
        raise HTTPException(status_code=404, detail="Document not found")
    if doc["tenantId"] != tenantId:
        raise HTTPException(status_code=403, detail="Access denied")
    
    if permanent:
        del documents_db[document_id]
    else:
        doc["deletedAt"] = datetime.utcnow().isoformat()
        doc["updatedAt"] = datetime.utcnow().isoformat()
    
    return {"success": True}

@router.post("/{document_id}/restore")
async def restore_document(
    document_id: str,
    tenantId: str = Form(...)
):
    """Restore a soft-deleted document"""
    doc = documents_db.get(document_id)
    if not doc:
        raise HTTPException(status_code=404, detail="Document not found")
    if doc["tenantId"] != tenantId:
        raise HTTPException(status_code=403, detail="Access denied")
    
    doc["deletedAt"] = None
    doc["updatedAt"] = datetime.utcnow().isoformat()
    return {k: v for k, v in doc.items() if k != "content"}

@router.post("/{document_id}/share")
async def share_document(
    document_id: str,
    tenantId: str = Form(...),
    expiresInHours: Optional[int] = Form(None)
):
    """Create a shareable link for a document"""
    doc = documents_db.get(document_id)
    if not doc:
        raise HTTPException(status_code=404, detail="Document not found")
    if doc["tenantId"] != tenantId:
        raise HTTPException(status_code=403, detail="Access denied")
    
    token = str(uuid.uuid4())
    doc["sharedLinkToken"] = token
    
    if expiresInHours:
        doc["sharedLinkExpiresAt"] = (datetime.utcnow() + timedelta(hours=expiresInHours)).isoformat()
    else:
        doc["sharedLinkExpiresAt"] = None
    
    doc["isShared"] = True
    doc["updatedAt"] = datetime.utcnow().isoformat()
    
    return {
        "token": token,
        "url": f"/api/share/{token}"
    }

@router.get("/{document_id}/breadcrumbs")
async def get_breadcrumbs(
    document_id: str,
    tenantId: str = Query(...)
):
    """Get breadcrumb path for a folder"""
    doc = documents_db.get(document_id)
    if not doc:
        raise HTTPException(status_code=404, detail="Document not found")
    if doc["tenantId"] != tenantId:
        raise HTTPException(status_code=403, detail="Access denied")
    
    breadcrumbs = [{"id": doc["id"], "name": doc["name"]}]
    
    current_id = doc.get("parentId")
    while current_id:
        parent = documents_db.get(current_id)
        if not parent:
            break
        breadcrumbs.insert(0, {"id": parent["id"], "name": parent["name"]})
        current_id = parent.get("parentId")
    
    breadcrumbs.insert(0, {"id": None, "name": "Home"})
    return breadcrumbs

@router.get("/share/{token}")
async def access_shared_document(token: str):
    """Access a shared document by token"""
    for doc in documents_db.values():
        if doc.get("sharedLinkToken") == token:
            # Check if expired
            expires_at = doc.get("sharedLinkExpiresAt")
            if expires_at:
                if datetime.utcnow() > datetime.fromisoformat(expires_at):
                    raise HTTPException(status_code=410, detail="Share link has expired")
            
            if doc["type"] == "file":
                content = doc.get("content", b"")
                return StreamingResponse(
                    iter([content]),
                    media_type=doc["mimeType"] or "application/octet-stream",
                    headers={"Content-Disposition": f'inline; filename="{doc["name"]}"'}
                )
            else:
                return {"id": doc["id"], "name": doc["name"], "type": "folder"}
    
    raise HTTPException(status_code=404, detail="Share link not found")

File: apps/backend/src/routes/documents.ts
import express from 'express';
import multer from 'multer';
import { v4 as uuidv4 } from 'uuid';

const router: express.Router = express.Router();
const upload = multer({ storage: multer.memoryStorage() });

// In-memory storage for demo (replace with database in production)
const documentsDb: Map<string, any> = new Map();

// Helper to get tenant from query
const getTenantId = (req: express.Request): string => {
  return (req.query.tenantId as string) || 'default-tenant';
};

// List documents
router.get('/', (req: express.Request, res: express.Response) => {
  const tenantId = getTenantId(req);
  const parentId = req.query.parentId as string | null;
  
  const docs = Array.from(documentsDb.values()).filter(
    (doc) => doc.tenantId === tenantId && doc.parentId === parentId && !doc.deletedAt
  );
  
  res.json(docs);
});

// Create folder
router.post('/folder', (req: express.Request, res: express.Response) => {
  const { name, parentId, tenantId } = req.body;
  const now = new Date().toISOString();
  
  const docId = uuidv4();
  const document = {
    id: docId,
    tenantId,
    parentId: parentId || null,
    name,
    type: 'folder',
    mimeType: null,
    sizeBytes: null,
    storagePath: null,
    isShared: false,
    sharedLinkToken: null,
    sharedLinkExpiresAt: null,
    createdBy: null,
    createdAt: now,
    updatedAt: now,
    deletedAt: null,
  };
  
  documentsDb.set(docId, document);
  res.json(document);
});

// Upload file
router.post('/upload', upload.single('file'), (req: express.Request, res: express.Response) => {
  const { parentId, tenantId } = req.body;
  const file = req.file;
  
  if (!file) {
    return res.status(400).json({ error: 'No file provided' });
  }
  
  const now = new Date().toISOString();
  const docId = uuidv4();
  
  const document = {
    id: docId,
    tenantId,
    parentId: parentId || null,
    name: file.originalname,
    type: 'file',
    mimeType: file.mimetype,
    sizeBytes: file.size,
    storagePath: `${tenantId}/${docId}`,
    isShared: false,
    sharedLinkToken: null,
    sharedLinkExpiresAt: null,
    createdBy: null,
    createdAt: now,
    updatedAt: now,
    deletedAt: null,
    content: file.buffer,
  };
  
  documentsDb.set(docId, document);
  
  const { content, ...responseDoc } = document;
  res.json(responseDoc);
});

// Download file
router.get('/:documentId/download', (req: express.Request, res: express.Response) => {
  const { documentId } = req.params;
  const tenantId = getTenantId(req);
  
  const doc = documentsDb.get(documentId);
  if (!doc || doc.tenantId !== tenantId) {
    return res.status(404).json({ error: 'Document not found' });
  }
  
  if (doc.type !== 'file') {
    return res.status(400).json({ error: 'Not a file' });
  }
  
  res.setHeader('Content-Disposition', `attachment; filename="${doc.name}"`);
  res.setHeader('Content-Type', doc.mimeType || 'application/octet-stream');
  res.send(doc.content);
});

// Preview file
router.get('/:documentId/preview', (req: express.Request, res: express.Response) => {
  const { documentId } = req.params;
  const tenantId = getTenantId(req);
  
  const doc = documentsDb.get(documentId);
  if (!doc || doc.tenantId !== tenantId) {
    return res.status(404).json({ error: 'Document not found' });
  }
  
  if (doc.type !== 'file') {
    return res.status(400).json({ error: 'Not a file' });
  }
  
  res.setHeader('Content-Disposition', `inline; filename="${doc.name}"`);
  res.setHeader('Content-Type', doc.mimeType || 'application/octet-stream');
  res.send(doc.content);
});

// Rename document
router.patch('/:documentId', (req: express.Request, res: express.Response) => {
  const { documentId } = req.params;
  const { name, tenantId } = req.body;
  
  const doc = documentsDb.get(documentId);
  if (!doc || doc.tenantId !== tenantId) {
    return res.status(404).json({ error: 'Document not found' });
  }
  
  doc.name = name;
  doc.updatedAt = new Date().toISOString();
  
  const { content, ...responseDoc } = doc;
  res.json(responseDoc);
});

// Move document
router.patch('/:documentId/move', (req: express.Request, res: express.Response) => {
  const { documentId } = req.params;
  const { parentId, tenantId } = req.body;
  
  const doc = documentsDb.get(documentId);
  if (!doc || doc.tenantId !== tenantId) {
    return res.status(404).json({ error: 'Document not found' });
  }
  
  doc.parentId = parentId || null;
  doc.updatedAt = new Date().toISOString();
  
  const { content, ...responseDoc } = doc;
  res.json(responseDoc);
});

// Delete document
router.delete('/:documentId', (req: express.Request, res: express.Response) => {
  const { documentId } = req.params;
  const tenantId = getTenantId(req);
  const permanent = req.query.permanent === 'true';
  
  const doc = documentsDb.get(documentId);
  if (!doc || doc.tenantId !== tenantId) {
    return res.status(404).json({ error: 'Document not found' });
  }
  
  if (permanent) {
    documentsDb.delete(documentId);
  } else {
    doc.deletedAt = new Date().toISOString();
    doc.updatedAt = new Date().toISOString();
  }
  
  res.json({ success: true });
});

// Restore document
router.post('/:documentId/restore', (req: express.Request, res: express.Response) => {
  const { documentId } = req.params;
  const { tenantId } = req.body;
  
  const doc = documentsDb.get(documentId);
  if (!doc || doc.tenantId !== tenantId) {
    return res.status(404).json({ error: 'Document not found' });
  }
  
  doc.deletedAt = null;
  doc.updatedAt = new Date().toISOString();
  
  const { content, ...responseDoc } = doc;
  res.json(responseDoc);
});

// Share document
router.post('/:documentId/share', (req: express.Request, res: express.Response) => {
  const { documentId } = req.params;
  const { tenantId, expiresInHours } = req.body;
  
  const doc = documentsDb.get(documentId);
  if (!doc || doc.tenantId !== tenantId) {
    return res.status(404).json({ error: 'Document not found' });
  }
  
  const token = uuidv4();
  doc.sharedLinkToken = token;
  
  if (expiresInHours) {
    const expiresAt = new Date();
    expiresAt.setHours(expiresAt.getHours() + parseInt(expiresInHours));
    doc.sharedLinkExpiresAt = expiresAt.toISOString();
  } else {
    doc.sharedLinkExpiresAt = null;
  }
  
  doc.isShared = true;
  doc.updatedAt = new Date().toISOString();
  
  res.json({
    token,
    url: `/api/share/${token}`,
  });
});

// Get breadcrumbs
router.get('/:documentId/breadcrumbs', (req: express.Request, res: express.Response) => {
  const { documentId } = req.params;
  const tenantId = getTenantId(req);
  
  const doc = documentsDb.get(documentId);
  if (!doc || doc.tenantId !== tenantId) {
    return res.status(404).json({ error: 'Document not found' });
  }
  
  const breadcrumbs: Array<{ id: string | null; name: string }> = [{ id: doc.id, name: doc.name }];
  
  let currentId = doc.parentId;
  while (currentId) {
    const parent = documentsDb.get(currentId);
    if (!parent) break;
    breadcrumbs.unshift({ id: parent.id, name: parent.name });
    currentId = parent.parentId;
  }
  
  breadcrumbs.unshift({ id: null, name: 'Home' });
  res.json(breadcrumbs);
});

// Access shared document
router.get('/share/:token', (req: express.Request, res: express.Response) => {
  const { token } = req.params;
  
  for (const doc of documentsDb.values()) {
    if (doc.sharedLinkToken === token) {
      if (doc.sharedLinkExpiresAt) {
        const expiresAt = new Date(doc.sharedLinkExpiresAt);
        if (new Date() > expiresAt) {
          return res.status(410).json({ error: 'Share link has expired' });
        }
      }
      
      if (doc.type === 'file') {
        res.setHeader('Content-Disposition', `inline; filename="${doc.name}"`);
        res.setHeader('Content-Type', doc.mimeType || 'application/octet-stream');
        return res.send(doc.content);
      } else {
        return res.json({ id: doc.id, name: doc.name, type: 'folder' });
      }
    }
  }
  
  res.status(404).json({ error: 'Share link not found' });
});

export default router;
